extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java

```diff
@@ -33,7 +36,33 @@ public FlywayCreator withCallbacks(Collection<Callback> callbacks) {
 
     public Flyway createFlyway(DataSource dataSource) {
         FluentConfiguration configure = Flyway.configure();
-        configure.dataSource(dataSource);
+
+        if (flywayRuntimeConfig.jdbcUrl.isPresent()) {
+            if (flywayRuntimeConfig.username.isPresent() && flywayRuntimeConfig.password.isPresent()) {
+                configure.dataSource(flywayRuntimeConfig.jdbcUrl.get(), flywayRuntimeConfig.username.get(),
+                        flywayRuntimeConfig.password.get());
+            } else {
+                throw new RuntimeException("User name and password must be set when the JDBC URL is provided");
+            }
+        } else {
+            if (flywayRuntimeConfig.username.isPresent() && flywayRuntimeConfig.password.isPresent()) {
+                AgroalDataSource agroalDataSource;
+                try {
+                    agroalDataSource = dataSource.unwrap(AgroalDataSource.class);
+                } catch (SQLException e) {
+                    // Should never happen
+                    throw new RuntimeException("Unable to unwrap " + dataSource + " to AgroalDataSource", e);
+                }
```

Comment:
The `unwrap` method in `io.agroal.pool.DataSource` always returns null. Better cast directly: 

```suggestion
                AgroalDataSource agroalDataSource = (AgroalDataSource) dataSource;
```